### PR TITLE
Enable NPS survey and migrate it to Foorm

### PR DESCRIPTION
**Pull request by jmkulwik**
This migrates the NPS survey to our foorm system and enables it. We plan to have it enabled for the ~2 week period from now until HOC begins.

Previously, the NPS survey was designed in HAML https://github.com/code-dot-org/code-dot-org/pull/27673/
This PR migrates it to our new foorm system. The JSON for the survey was created using [surveyjs](https://surveyjs.io/create-survey#). The default survey-library css overrides can be found here: https://github.com/surveyjs/survey-library/tree/master/src/defaultCss

NPS score results will be reported to Google Analytics under survey/nps2020

Full survey results will be reported to the table foorm_submissions with the filter form_name = surveys/teachers/nps_survey and form_version = 0
`SELECT * FROM foorm_submissions WHERE form_name='surveys/teachers/nps_survey' AND form_version = 0;`

In context images:
image

image

image

Error handling has been added as well:
image

image

image

image

## Links

- [spec](https://docs.google.com/document/d/1g3EUEizzRveh6FBH2nOM-k7aqmTvr-zUJ72VPJDhnlo/edit?ts=5fa1de80#)
- [LP-1636](https://codedotorg.atlassian.net/browse/LP-1636)

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/config/foorm/forms/surveys/teachers/nps_survey.0.json [hunk 1]
```diff
@@ -0,0 +1,97 @@
+{
```
Comment (jmkulwik) [inline dashboard/config/foorm/forms/surveys/teachers/nps_survey.0.json:1]: Generated json from https://surveyjs.io/create-survey#

File: dashboard/app/controllers/foorm/misc_survey_controller.rb [hunk 2]
```diff
@@ -45,6 +45,31 @@ def new
       }
     end
 
+    def show
```
Comment (molly-moen) [inline dashboard/app/controllers/foorm/misc_survey_controller.rb:53]: can you add a comment to this method? So in the future we know the difference between show and new

File: dashboard/config/foorm/forms/surveys/teachers/nps_survey.0.json [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      "visibleIf": "{nps_value} notempty",
+      "title": "What is the top area of our website and curriculum that we should focus on improving?",
+      "isRequired": true,
+      "requiredErrorText": "Please select up to three areas",
+      "validators": [
+       {
+        "type": "answercount",
+        "text": "Please select a maximum of three areas",
+        "minCount": 1,
+        "maxCount": 3
```
Comment (molly-moen) [inline dashboard/config/foorm/forms/surveys/teachers/nps_survey.0.json:28]: TIL you can set a min/max count on checkbox responses!

File: dashboard/app/controllers/foorm/misc_survey_controller.rb (hunk 2 above)
Comment (bencodeorg) [reply dashboard/app/controllers/foorm/misc_survey_controller.rb:53]: Semi-related nit -- maybe route could be `/form/:misc_form_path/show`?

File: apps/src/code-studio/pd/foorm/Foorm.jsx [hunk 4]
```diff
@@ -19,10 +19,16 @@ export default class Foorm extends React.Component {
...
     submitApi: PropTypes.string.isRequired,
     surveyData: PropTypes.object,
-    submitParams: PropTypes.object
+    submitParams: PropTypes.object,
+    customCssClasses: PropTypes.object,
+    onComplete: PropTypes.func
   };
 
-  customCss = {
+  static defaultProps = {
```
Comment (bencodeorg) [inline apps/src/code-studio/pd/foorm/Foorm.jsx:27]: TIL about defaultProps!

**Review by jamescodeorg — COMMENTED** [review]
Mostly curious, is this the typical way we will incorporate a survey into our site (with a new React component and custom styling)?  Or was there something unusual about the requirements that required us to do additional work?  (I was kind of expecting just a survey definition and a banner/link to pre-made survey UI somewhere.)

**Comment by jmkulwik** [comment]
> Mostly curious, is this the typical way we will incorporate a survey into our site (with a new React component and custom styling)? Or was there something unusual about the requirements that required us to do additional work? (I was kind of expecting just a survey definition and a banner/link to pre-made survey UI somewhere.)

The short answer: The default survey UI is very different from the teacher homepage UI. I initially just had the survey show up with default styling and it looked pretty odd. On top of that, due to the homepage css, the styling made the survey almost impossible to fill out. AFAIK (without looking into it) we only have one other survey like this, so a one-off didn't feel wrong.

The long answer: The vast majority of our surveys are much larger and are presented after a PD. The users are asked to fill them out as part of the PD and are given time to do so. They navigate to a page where the survey lives and fill out multiple pages of questions. 

This one's a bit different because it's essentially a 1 question survey that we want to be as easy to fill out as possible. We want it to show up as part of the page and feel like it's part of the site.

File: dashboard/app/controllers/foorm/misc_survey_controller.rb [hunk 5]
```diff
@@ -45,6 +45,11 @@ def new
       }
     end
 
+    # Gets a json format of a general misc survey.
+    # GET '/form_data/:misc_form_path'
```
Comment (molly-moen) [inline dashboard/app/controllers/foorm/misc_survey_controller.rb]: update path here?

File: dashboard/test/controllers/home_controller_test.rb [hunk 6]
```diff
@@ -362,7 +362,7 @@ class HomeControllerTest < ActionController::TestCase
   # TODO: remove this test when workshop_organizer is deprecated
   test 'workshop organizers see dashboard links' do
     sign_in create(:workshop_organizer, :with_terms_of_service)
-    query_count = 15
+    query_count = 14
```
Comment (molly-moen) [inline dashboard/test/controllers/home_controller_test.rb:365]: out of curiosity: did this change because of the removal of the nps survey in haml?

Comment (jmkulwik) [reply dashboard/test/controllers/home_controller_test.rb:365]: Yes. Specifically the removal of `show_nps_survey?` in `dashboard/app/views/home/_homepage.html.haml`

File: dashboard/app/controllers/foorm/misc_survey_controller.rb (hunk 5 above)
Comment (jmkulwik) [reply dashboard/app/controllers/foorm/misc_survey_controller.rb]: Good catch. Thank you!

File: dashboard/app/controllers/foorm/misc_survey_controller.rb (hunk 2 above)
Comment (jmkulwik) [reply dashboard/app/controllers/foorm/misc_survey_controller.rb:53]: 👍

**Review by molly-moen — APPROVED** [review]
This is awesome!

File: dashboard/app/controllers/home_controller.rb [hunk 7]
```diff
@@ -147,6 +148,7 @@ def init_homepage
       @homepage_data[:announcement] = DCDO.get('announcement_override', nil)
       @homepage_data[:hiddenScripts] = current_user.get_hidden_script_ids
       @homepage_data[:showCensusBanner] = show_census_banner
+      @homepage_data[:showNpsSurvey] = show_nps_survey? SurveyResult::NET_PROMOTER_SCORE_2020
```
Comment (bencodeorg) [inline dashboard/app/controllers/home_controller.rb:151]: Is this constant doing anything? Seems like it was used to store results in `SurveyResults`, but we're not storing results there any more?

Comment (bencodeorg) [reply dashboard/app/controllers/home_controller.rb:151]: I think maybe having that constant stored in `KINDS` in SurveyResults implies there should be results for that type of survey stored there, which there won't be I don't think?

**Review by bencodeorg — APPROVED** [review]
Looks great! Cool to see foorm out in the wild.

**Comment by breville** [comment]
Very cool to see Foorm used more broadly.  There are some good questions here about what we'll need to do each time we adapt Foorm to a new UI.  We used SurveyJS for a fast lift on the form UI, however we also considered writing our own form UI instead.  If we end up using Foorm in a variety of situations which have significantly-different styling, we might consider that again.  Given that the CSS-based styling we use isn't the most elegant solution in a React world, we might want to make that decision sooner rather than later.

**Pull request merged**